### PR TITLE
Realization of hashtable + corresponding tests.

### PR DESCRIPTION
**Pull request by gorshkovayulia**
@ctapobep please take a look.

### CODE REVIEW

File: src/main/java/hashtable/HashTable.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        return node == null ? null : node.value;
+    }
+
+    /**
+     * Returns the index in table for specified key.
+     * @param key the key that needs to be converted into index.
+     * @return index in table.
+     * @exception IllegalArgumentException if specified key is null.
+     */
+    public int getIndexForKey(K key) {
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable.java:87]: Ideally the number of public methods is kept to the minimum. Methods like this shouldn't be called from outside, so it can be private.

File: src/main/java/hashtable/LinkedListNode.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package hashtable;
+
+public class LinkedListNode<K, V> {
+    public LinkedListNode<K, V> next;
```
Comment (ctapobep) [inline src/main/java/hashtable/LinkedListNode.java:4]: Hm.. public field?

File: src/test/java/hashtable/HashTableTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
+package hashtable;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class HashTableTest {
+
+    @Test(expected = IllegalArgumentException.class)
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:9]: Latest JUnit (4.13.1) has an assert which allows this: 

```
Exception e = assertThrows(IllegalArgumentException.class, () -> new HashTable(-1));
assertEquals("Capacity cannot be less or equal zero!", e.getMessage());
```

File: src/test/java/hashtable/HashTableTest.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void impossibleToSetZeroCapacity() {
+        new HashTable(0);
+    }
+
+    @Test
+    public void possibleToSetPositiveCapacity() {
+        new HashTable(2);
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:21]: The boundary value for this test is 1 ;)

File: src/test/java/hashtable/HashTableTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test
+    public void possibleToPutElementInEmptySlot() {
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:25]: We shouldn't test private parts of the algorithm. We have public API like `getValue()`, `add()`, `remove()` - these are the methods that we can use for testing.

File: src/test/java/hashtable/HashTableTest.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void impossibleToRemoveElementWithNullKey() {
+        HashTable table = new HashTable(10);
+        table.remove(null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void impossibleToRemoveNotExistingElement() {
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:54]: I'd simply do nothing in such scenario instead of throwing an exception.

File: src/test/java/hashtable/HashTableTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void impossibleToGetValueForNullKey() {
+        HashTable table = new HashTable(10);
+        table.getValue(null);
+    }
+
+    @Test
+    public void nullValueIsReturnedForEmptyIndex() {
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:92]: It's not an index.. Maybe something like `returnsNullValueForNonExistingKey`

File: src/test/java/hashtable/HashTableTest.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    @Test
+    public void valueOfTenIsReturnedForTheFirstIndex() {
+        HashTable table = new HashTable(10);
+        table.put(0, 10);
+        table.put(1, 20);
+        assertEquals(10, table.getValue(0));
+    }
+
+    @Test
+    public void valueOfFifteenIsReturnedForTheMiddleIndex() {
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:106]: `returnsValuesAddedBeforeLast`?

File: src/test/java/hashtable/HashTableTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test
+    public void possibleToGetIndexForNumberKey() {
+        HashTable table = new HashTable(10);
+        assertEquals(6, table.getIndexForKey(6));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void impossibleToGetIndexForNullKey() {
```
Comment (ctapobep) [inline src/test/java/hashtable/HashTableTest.java:135]: This should probably return null. We can treat it as if someone requested a value for a non-existing (null) key.
